### PR TITLE
UIU-1279: Apply Custom Fields page components for record

### PR DESCRIPTION
**Pull request by Vladyslav-Velytskyi**
#### Create `ViewCustomFieldsRecord` accordion
- It's implement accordion on the View Record
- Accordion Label takes from `Settings > MODULE_NAME > Custom fields > Section title value`
- Display custom fields where Hidden = No
- Grid display in the same manner as other accordions that contain checkboxes, text boxes, single select and multi select
- If no value set then shows `<NoValue />` component
#### Create `EditCustomFieldsRecord` accordion
- It's implement accordion on the Create/Edit Record
- Accordion Label takes from `Settings > MODULE_NAME > Custom fields > Section title value`
- Display custom fields where Hidden = No
- Grid display in the same manner as other accordions that contain checkboxes, text boxes, single select and multi select
- Correctly works for `react-final-form` and `redux-form`
- Correclty works only for `<Fields />` where component is textarea or text box, 
- If a custom field has help text then display the info popover icon to the right of the field 
- Validation for required fields. Limit for text boxes 150 characters and fro textareas is 1500 characters

Example of connection in [`ui-users`](https://github.com/folio-org/ui-users/pull/1201)

## Issue
[UIU-1279](https://issues.folio.org/browse/UIU-1279)

### User Details Page
image

### User Edit Page
image

### CODE REVIEW

File: lib/CustomFields/pages/EditCustomFieldsRecord/EditCustomFieldsRecord.js [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    sectionTitleFetchFailed,
+  } = useSectionTitleFetch(okapi, backendModuleName.toUpperCase());
+  const { calloutRef } = useLoadingErrorCallout(customFieldsFetchFailed || sectionTitleFetchFailed);
+  const dataIsLoaded = sectionTitleLoaded && customFieldsLoaded;
+  const customFieldsIsVisible = customFields?.length > 0 && customFields?.some(customField => customField.visible);
+  const accordionIsVisible = dataIsLoaded && customFieldsIsVisible;
+
+  const Field = fieldComponent;
+  const formatedCustomFields = chunk(customFields, columnCount);
+  const columnWidth = 12 / columnCount;
```
Comment (BogdanDenis) [inline lib/CustomFields/pages/EditCustomFieldsRecord/EditCustomFieldsRecord.js]: Please move `12` to a named constant

File: lib/CustomFields/pages/ViewCustomFieldsRecord/ViewCustomFieldsRecord.js [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        ? (
+          <Accordion
+            open={expanded}
+            id={accordionId}
+            onToggle={onToggle}
+            label={sectionTitle.value || defaultLabel}
+          >
+            {
+              formatedCustomFields.map((row, i) => (
+                <Row key={i}>
```
Comment (BogdanDenis) [inline lib/CustomFields/pages/ViewCustomFieldsRecord/ViewCustomFieldsRecord.js]: Can we use ids or names from custom fields here as `key`?

File: lib/CustomFields/pages/ViewCustomFieldsRecord/ViewCustomFieldsRecord.js [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+                  {
+                    row.map(customField => (
+                      customField.visible ? (
+                        <Col
+                          key={customField.refId}
+                          xs={12 / columnCount}
```
Comment (BogdanDenis) [inline lib/CustomFields/pages/ViewCustomFieldsRecord/ViewCustomFieldsRecord.js]: Same here about column width

File: lib/CustomFields/pages/ViewCustomFieldsRecord/ViewCustomFieldsRecord.js (hunk 2 above)
Comment (Vladyslav-Velytskyi) [reply lib/CustomFields/pages/ViewCustomFieldsRecord/ViewCustomFieldsRecord.js]: No, cause `formattedCustomFields` it's just `customFields` which splitted in groups by `chunk()` from lodash. This is done in order to correctly display columns

File: lib/CustomFields/pages/EditCustomFieldsRecord/EditCustomFieldsRecord.js [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+  return (
+    <>
+      {accordionIsVisible
+        ? (
+          <Accordion
+            open={expanded}
+            id={accordionId}
+            onToggle={onToggle}
+            label={sectionTitle.value || defaultLabel}
```
Comment (MaksymDryha) [inline lib/CustomFields/pages/EditCustomFieldsRecord/EditCustomFieldsRecord.js]: I don't think we need to have a default label. The accordion won't be rendered if no section title is set

File: lib/CustomFields/pages/EditCustomFieldsRecord/EditCustomFieldsRecord.js [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+                                {customField.helpText && (
+                                  <InfoPopover
+                                    content={customField.helpText}
+                                    iconSize="small"
+                                  />
+                                )}
+                              </>
+                            }
+                            validate={value => {
+                              if (valuesValidation[customField.type]) {
```
Comment (MaksymDryha) [inline lib/CustomFields/pages/EditCustomFieldsRecord/EditCustomFieldsRecord.js]: I would suggest to make `valuesValidation` a curried function that accepts `type` and `value`. (Also, I'd better name it `validateCustomField`).Then here we'll be able to just write `validate={validateCustomField(customField.type)}`

File: lib/CustomFields/pages/EditCustomFieldsRecord/EditCustomFieldsRecord.js [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+  backendModuleName,
+  entityType,
+  fieldComponent,
+  columnCount,
+}) => {
+  const {
+    customFields,
+    customFieldsLoaded,
+    customFieldsFetchFailed,
+  } = useCustomFieldsFetch(okapi, backendModuleId, entityType);
```
Comment (MaksymDryha) [inline lib/CustomFields/pages/EditCustomFieldsRecord/EditCustomFieldsRecord.js]: Please, put an empty line here

File: lib/CustomFields/pages/EditCustomFieldsRecord/EditCustomFieldsRecord.js [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+  const {
+    sectionTitle,
+    sectionTitleLoaded,
+    sectionTitleFetchFailed,
+  } = useSectionTitleFetch(okapi, backendModuleName.toUpperCase());
```
Comment (MaksymDryha) [inline lib/CustomFields/pages/EditCustomFieldsRecord/EditCustomFieldsRecord.js]: Please, put an empty line here

File: lib/CustomFields/pages/EditCustomFieldsRecord/EditCustomFieldsRecord.js [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+                                  <InfoPopover
+                                    content={customField.helpText}
+                                    iconSize="small"
+                                  />
+                                )}
+                              </>
+                            }
+                            validate={validateCustomFields(customField)}
+                          />
+                        </Col>
```
Comment (MaksymDryha) [inline lib/CustomFields/pages/EditCustomFieldsRecord/EditCustomFieldsRecord.js]: I would suggest to move it to `renderCustomField` function, but it's not a strong preference. Also, there's no need for a ternary operator. We can simply write `customField.visible && renderCustomField(customField)`

File: lib/CustomFields/pages/EditCustomFieldsRecord/EditCustomFieldsRecord.js [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+import {
+  selectModuleId,
+  selectOkapiData,
+} from '../../selectors';
+import {
+  useCustomFieldsFetch,
+  useSectionTitleFetch,
+  useLoadingErrorCallout,
+  validateCustomFields,
+} from '../../utils';
```
Comment (MaksymDryha) [inline lib/CustomFields/pages/EditCustomFieldsRecord/EditCustomFieldsRecord.js]: Please, put an empty line after each multiline import

File: lib/CustomFields/pages/EditCustomFieldsRecord/EditCustomFieldsRecord.js [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+  const customFieldsIsVisible = customFields?.length && customFields?.some(customField => customField.visible);
+  const accordionIsVisible = dataIsLoaded && customFieldsIsVisible;
+  const Field = fieldComponent;
+  const formatedCustomFields = chunk(customFields, columnCount);
+  const columnWidth = rowShapes / columnCount;
+
+  return (
+    <>
+      {accordionIsVisible
+        ? (
```
Comment (MaksymDryha) [inline lib/CustomFields/pages/EditCustomFieldsRecord/EditCustomFieldsRecord.js]: We can replace the ternary operator with `accordionIsVisible && ...`

File: lib/CustomFields/pages/ViewCustomFieldsRecord/ViewCustomFieldsRecord.js [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+  NoValue,
+  Row,
+  Col,
+} from '@folio/stripes-components';
+
+import { rowShapes } from '../../constants';
+import {
+  selectModuleId,
+  selectOkapiData,
+} from '../../selectors';
```
Comment (MaksymDryha) [inline lib/CustomFields/pages/ViewCustomFieldsRecord/ViewCustomFieldsRecord.js]: Please, put an empty line here

File: lib/CustomFields/pages/ViewCustomFieldsRecord/ViewCustomFieldsRecord.js [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+  backendModuleName,
+  entityType,
+  customFieldsValues,
+  columnCount,
+}) => {
+  const {
+    customFields,
+    customFieldsLoaded,
+    customFieldsFetchFailed,
+  } = useCustomFieldsFetch(okapi, backendModuleId, entityType);
```
Comment (MaksymDryha) [inline lib/CustomFields/pages/ViewCustomFieldsRecord/ViewCustomFieldsRecord.js]: Please, put an empty line here

File: lib/CustomFields/pages/ViewCustomFieldsRecord/ViewCustomFieldsRecord.js [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+  const {
+    sectionTitle,
+    sectionTitleLoaded,
+    sectionTitleFetchFailed,
+  } = useSectionTitleFetch(okapi, backendModuleName.toUpperCase());
```
Comment (MaksymDryha) [inline lib/CustomFields/pages/ViewCustomFieldsRecord/ViewCustomFieldsRecord.js]: Please, put an empty line here

File: lib/CustomFields/pages/ViewCustomFieldsRecord/ViewCustomFieldsRecord.js [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+  const dataIsLoaded = sectionTitleLoaded && customFieldsLoaded;
+  const customFieldsIsVisible = customFields?.length && customFields?.some(customField => customField.visible);
+  const accordionIsVisible = dataIsLoaded && customFieldsIsVisible;
+  const columnWidth = rowShapes / columnCount;
+  const formatedCustomFields = chunk(customFields, columnCount);
+
+  return (
+    <>
+      {accordionIsVisible
+        ? (
```
Comment (MaksymDryha) [inline lib/CustomFields/pages/ViewCustomFieldsRecord/ViewCustomFieldsRecord.js]: The ternary operator is not necessary here

File: lib/CustomFields/pages/ViewCustomFieldsRecord/ViewCustomFieldsRecord.js [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+            id={accordionId}
+            onToggle={onToggle}
+            label={sectionTitle.value}
+          >
+            {
+              formatedCustomFields.map((row, i) => (
+                <Row key={i}>
+                  {
+                    row.map(customField => (
+                      customField.visible ? (
```
Comment (MaksymDryha) [inline lib/CustomFields/pages/ViewCustomFieldsRecord/ViewCustomFieldsRecord.js]: The same about ternary operator

File: lib/CustomFields/utils/validateCustomFields.js [hunk 16]
```diff
@@ -0,0 +1,23 @@
+import React from 'react';
+import { FormattedMessage } from 'react-intl';
+import { fieldLimits } from '../constants';
+
+const validateCustomFields = customField => value => {
```
Comment (MaksymDryha) [inline lib/CustomFields/utils/validateCustomFields.js]: Looks great

**Pull request merged**